include/ginkgo/core/base/utils_helper.hpp

```diff
@@ -211,6 +212,8 @@ inline detail::shared_type<OwningPointer> share(OwningPointer&& p)
 {
     static_assert(detail::have_ownership<OwningPointer>(),
                   "OwningPointer does not have ownership of the object");
+    static_assert(std::is_rvalue_reference<decltype(p)>::value,
+                  "p must be an xvalue for this function to work");
```

Comment:
coming from the caller perspective, p can be either a prvalue or an xvalue, so I would suggest adapting the error message to the caller instead of the callee.
```suggestion
    static_assert(std::is_rvalue_reference<decltype(p)>::value,
                  "p must be an rvalue for this function to work");
```